File: scripts/algorithms/M/Minimum Operations to Make the Array Alternating/Minimum Operations to Make the Array Alternating.py
# Runtime: 3233 ms (Top 5.15%) | Memory: 31.6 MB (Top 46.98%)
class Solution:
    def minimumOperations(self, nums: List[int]) -> int:
        n = len(nums)
        odd, even = defaultdict(int), defaultdict(int)
        for i in range(n):
            if i % 2 == 0:
                even[nums[i]] += 1
            else:
                odd[nums[i]] += 1
        topEven, secondEven = (None, 0), (None, 0)
        for num in even:
            if even[num] > topEven[1]:
                topEven, secondEven = (num, even[num]), topEven
            elif even[num] > secondEven[1]:
                secondEven = (num, even[num])
        topOdd, secondOdd = (None, 0), (None, 0)
        for num in odd:
            if odd[num] > topOdd[1]:
                topOdd, secondOdd = (num, odd[num]), topOdd
            elif odd[num] > secondOdd[1]:
                secondOdd = (num, odd[num])
        if topOdd[0] != topEven[0]:
            return n - topOdd[1] - topEven[1]
        else:
            return n - max(secondOdd[1] + topEven[1], secondEven[1] + topOdd[1])

File: scripts/algorithms/M/Minimum Operations to Make the Array Alternating/Minimum Operations to Make the Array Alternating.cpp
// Runtime: 775 ms (Top 27.44%) | Memory: 140.3 MB (Top 88.77%)
class Solution {
public:
    int minimumOperations(vector<int>& nums) {

        int totalEven = 0, totalOdd = 0;

        unordered_map<int,int> mapEven, mapOdd;

        for(int i=0;i<nums.size();i++) {
            if(i%2==0) {
                totalEven++;
                mapEven[nums[i]]++;
            }

            else {
                totalOdd++;
                mapOdd[nums[i]]++;
            }
        }

        int firstEvenCount = 0, firstEven = 0;
        int secondEvenCount = 0, secondEven = 0;

        for(auto it=mapEven.begin();it!=mapEven.end();it++) {
            int num = it->first;
            int count = it->second;

            if(count>=firstEvenCount) {
                secondEvenCount = firstEvenCount;
                secondEven = firstEven;
                firstEvenCount = count;
                firstEven = num;
            }

            else if(count >= secondEvenCount) {
                secondEvenCount = count;
                secondEven = num;
            }
        }

        int firstOddCount = 0, firstOdd = 0;
        int secondOddCount = 0, secondOdd = 0;

        for(auto it=mapOdd.begin();it!=mapOdd.end();it++) {
            int num = it->first;
            int count = it->second;

            if(count>=firstOddCount) {
                secondOddCount = firstOddCount;
                secondOdd = firstOdd;
                firstOddCount = count;
                firstOdd = num;
            }

            else if(count>=secondOddCount) {
                secondOddCount = count;
                secondOdd = num;
            }
        }

        int operationsEven = 0, operationsOdd = 0;

        operationsEven = totalEven - firstEvenCount;

        if(firstEven!=firstOdd) operationsEven += (totalOdd - firstOddCount);
        else operationsEven += (totalOdd - secondOddCount);

        operationsOdd = totalOdd - firstOddCount;
        if(firstOdd!=firstEven) operationsOdd += (totalEven - firstEvenCount);
        else operationsOdd += (totalEven - secondEvenCount);

        return min(operationsEven, operationsOdd);

    }
};